##**Student name: Ankita Kandalkar**

##**Hamoye ID: 147a35c062c1f000**

In [44]:
from google.colab import files

uploaded = files.upload()

Saving energydata_complete.csv to energydata_complete (2).csv


In [48]:
import pandas as pd
import io
  
df = pd.read_csv(io.BytesIO(uploaded['energydata_complete.csv']),encoding='latin1')
#print(df)

pd.set_option('display.max_columns', None)
pd.reset_option('max_columns')

df.head()

date  Appliances  lights  ...  Tdewpoint        rv1        rv2
0  2016-01-11 17:00:00          60      30  ...        5.3  13.275433  13.275433
1  2016-01-11 17:10:00          60      30  ...        5.2  18.606195  18.606195
2  2016-01-11 17:20:00          50      30  ...        5.1  28.642668  28.642668
3  2016-01-11 17:30:00          50      40  ...        5.0  45.410389  45.410389
4  2016-01-11 17:40:00          60      40  ...        4.9  10.084097  10.084097

[5 rows x 29 columns]

________________________________________________________________________________

### **Question no. 12:** 

**From the dataset, fit a linear model on the relationship between the temperature in the living room in Celsius (x = T2) and the temperature outside the building (y = T6). What is the R^2 value in two d.p.?**

In [49]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.model_selection import train_test_split

In [50]:
X = df["T2"].values
Y = df['T6'].values
print(X,"\n",Y)

[19.2        19.2        19.2        ... 25.62857143 25.414
 25.26428571] 
 [ 7.02666667  6.83333333  6.56       ... 23.62666667 22.43333333
 21.02666667]


In [52]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42)
print(X_train.shape, Y_train.shape)
print(X_test.shape, Y_test.shape)

(15788,) (15788,)
(3947,) (3947,)


In [53]:
reg = LinearRegression()
reg.fit(X_train.reshape(-1, 1) ,Y_train)
print(reg.coef_)

[2.22739928]


In [54]:
Y_pred = reg.predict(X_test.reshape(-1, 1))

In [55]:
print("R score is: ", round(r2_score(Y_test, Y_pred),2))

R score is:  0.64


**Answer of question no. 12** 

*R^2 value = 0.64*

________________________________________________________________________________

### **Question no. 13:** 

**Normalize the dataset using the MinMaxScaler after removing the following columns: [“date”, “lights”]. The target variable is “Appliances”. Use a 70-30 train-test set split with a random state of 42 (for reproducibility). Run a multiple linear regression using the training set and evaluate your model on the test set.** 

**What is the Mean Absolute Error (in two decimal places)?**


In [12]:
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error

In [56]:
new_df = df.drop(columns= ["date", "lights"])
new_df.head()

Appliances     T1       RH_1  ...  Tdewpoint        rv1        rv2
0          60  19.89  47.596667  ...        5.3  13.275433  13.275433
1          60  19.89  46.693333  ...        5.2  18.606195  18.606195
2          50  19.89  46.300000  ...        5.1  28.642668  28.642668
3          50  19.89  46.066667  ...        5.0  45.410389  45.410389
4          60  19.89  46.333333  ...        4.9  10.084097  10.084097

[5 rows x 27 columns]

In [57]:
X = new_df.drop(columns=["Appliances"]).values
Y = new_df["Appliances"].values
print(X,"\n\n",Y)

[[19.89       47.59666667 19.2        ...  5.3        13.27543316
  13.27543316]
 [19.89       46.69333333 19.2        ...  5.2        18.60619498
  18.60619498]
 [19.89       46.3        19.2        ...  5.1        28.64266817
  28.64266817]
 ...
 [25.5        46.59666667 25.62857143 ... 13.26666667 29.19911708
  29.19911708]
 [25.5        46.99       25.414      ... 13.23333333  6.32278365
   6.32278365]
 [25.5        46.6        25.26428571 ... 13.2        34.11885059
  34.11885059]] 

 [ 60  60  50 ... 270 420 430]


In [58]:
scale_X = MinMaxScaler()
X = scale_X.fit_transform(X)

In [59]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3, random_state=42)
print(X_train.shape, Y_train.shape)
print(X_test.shape, Y_test.shape)

(13814, 26) (13814,)
(5921, 26) (5921,)


In [60]:
reg = LinearRegression()
reg.fit(X_train ,Y_train)
print(reg.coef_)

[  -3.51072478  592.29486183 -252.710373   -488.66680472  310.97103712
  102.77165244   31.00966845   28.23278127  -16.75282177   17.12619362
  252.97465249   40.7120577    11.04109074  -47.73659198  109.13469841
 -168.62716102 -203.23662712  -42.58633878 -344.3898449     7.31807923
  -83.1075989    31.22594824   13.16807161  126.00077634    0.824005
    0.824005  ]


In [61]:
Y_pred = reg.predict(X_test)

In [62]:
print("Mean absolute error is: ", round(mean_absolute_error(Y_test, Y_pred),2))

Mean absolute error is:  53.64


**Answer of question no. 13** 

*Mean absolute error is:  53.64* 

________________________________________________________________________________

###**Question no. 14:**

**What is the Residual Sum of Squares (in two decimal places)?**

In [66]:
print("Residual sum of squares: %.2f" % ((Y_pred - Y_test) ** 2).sum())

Residual sum of squares: 51918501.21


**Answer of question no. 14**

*Residual sum of squares: 51918501.21*

________________________________________________________________________________

###**Question no. 15:** 

**What is the Root Mean Squared Error (in three decimal places)?**

In [22]:
from sklearn.metrics import mean_squared_error

In [24]:
RMS = mean_squared_error(Y_test, Y_pred, squared=False)
print(RMS)
print("Root Mean Square Error: ",round(RMS,3))

93.64046093998029
Root Mean Square Error:  93.64


**Answer of question no. 15** 

*Root Mean Square Error:  93.640*


________________________________________________________________________________

###**Question no. 16:** 

**What is the Coefficient of Determination (in two decimal places)?**

In [64]:
from sklearn.metrics import r2_score

In [65]:
print("Value of Coefficient of Determination: ", round(r2_score(Y_test,Y_pred),2))

Value of Coefficient of Determination:  0.15


**Answer of question no. 16** 

*Value of Coefficient of Determination :  0.15*

________________________________________________________________________________

###**Question no. 17:** 

**Obtain the feature weights from your linear model above. Which features have the lowest and highest weights respectively?**

In [27]:
print(list(reg.coef_).index(min(reg.coef_)))
print(list(reg.coef_).index(max(reg.coef_)))
print(reg.coef_)

3
1
[  -3.51072478  592.29486183 -252.710373   -488.66680472  310.97103712
  102.77165244   31.00966845   28.23278127  -16.75282177   17.12619362
  252.97465249   40.7120577    11.04109074  -47.73659198  109.13469841
 -168.62716102 -203.23662712  -42.58633878 -344.3898449     7.31807923
  -83.1075989    31.22594824   13.16807161  126.00077634    0.824005
    0.824005  ]


In [40]:
print("Minimum Coefficient value: ", new_df.columns[list(reg.coef_).index(min(reg.coef_))+1]," ", round(min(reg.coef_),3))
print("Maximum Coefficient value: ", new_df.columns[list(reg.coef_).index(max(reg.coef_))+1]," ", round(max(reg.coef_),3))

Minimum Coefficient value:  RH_2   -488.667
Maximum Coefficient value:  RH_1   592.295


**Answer of question no. 17** 

*RH_2 features has Minimum Coefficient value:  -488.667*

*RH_1 feature has Maximum Coefficient value:  592.295*

________________________________________________________________________________

###**Question no. 18:** 

**Train a ridge regression model with an alpha value of 0.4. Is there any change to the root mean squared error (RMSE) when evaluated on the test set?**

In [29]:
from sklearn.linear_model import Ridge

In [30]:
clf = Ridge(alpha=0.4)
clf.fit(X_train, Y_train)

Ridge(alpha=0.4)

In [31]:
Y_pred_ridge = clf.predict(X_test)

In [32]:
RMS = mean_squared_error(Y_test, Y_pred_ridge, squared=False)
print("Root Mean Square Error: ",round(RMS,3))

Root Mean Square Error:  93.661


**Answer of question no. 18** 

*Root Mean Square Error:  93.661*

________________________________________________________________________________

###**Question no. 19:** 

**Train a lasso regression model with an alpha value of 0.001 and obtain the new feature weights with it. How many of the features have non-zero feature weights?**

In [33]:
from sklearn import linear_model

In [34]:
clf = linear_model.Lasso(alpha=0.001)
clf.fit(X_train,Y_train)

Lasso(alpha=0.001)

In [35]:
Y_pred_lasso = clf.predict(X_test)

In [36]:
non_zero_coeff = 0
for coef in clf.coef_:
   if(coef!=0.0):
     non_zero_coeff+=1

print("Non Zero Coeffs: ", non_zero_coeff)
print(clf.coef_)

Non Zero Coeffs:  25
[  -4.14712394  590.59982716 -249.46879771 -484.56533287  310.21038378
  102.20461099   30.34519749   27.44590464  -17.11872329   17.10332116
  250.08097851   40.09288319   10.34680867  -47.38909406  108.88342254
 -168.69547613 -202.5025445   -42.45510038 -332.57169255    7.28215443
  -78.00038519   31.38951445   13.07831882  118.5980301     1.63117421
    0.        ]


**Answer of question no. 19**

*Number of Non Zero Coefficients: 25*

___________________________________________________________________________ 

###**Question no. 20:** 

**What is the new RMSE with the lasso regression? (Answer should be in three (3) decimal places)**

In [38]:
RMS = mean_squared_error(Y_test, Y_pred_lasso, squared=False)
print("Root Mean Square Error: ",round(RMS,3))

Root Mean Square Error:  93.641


**Answer of question no. 20**

*Root Mean Square Error:  93.641*

________________________________________________________________________________